### DISCUSSION TITLE
Feature flags 🔥

### DISCUSSION THREAD

**Discussion by arsangamal** [keycloak/keycloak#29711] [Ideas]
# Feature flags
a software development concept that allow you to enable or disable a feature without modifying the source code or requiring a redeploy.

## Introduction
As Keycloak acts as the main system for all organizations that operate and depend mainly on it, I think one of the really needed features is feature flags. Centralizing the control of many aspects of the system from one place is very important, specially for SaaS and Multi-tenant projects, which Keycloak already fits in perfectly.

## Pros
- Adding Feature flags to Keycloak will logically add an authorization layer above, so you won't have many roundtrips between the application and Keycloak for checking whether a user has the permission or not.
- Feature flags can be applied on all levels
  - Keycloak level
  - Realm Level
  - Client Level
  - Group Level
  - User Level

- Feature flags can be a one level switch or nested, this will help fit all kinds of systems from simple to complex ones
- Feature flags might have dependencies, so for example if feature X is enabled, feature Y is enabled/available.

This is a simple idea, if accepted I think it will have more and more ideas from the community
Thanks

**Comment by merlosy** [comment]
How about integrating Keycloak with Unleash? https://github.com/Unleash/unleash?tab=readme-ov-file
We could have the activated features for a specific user, let's say in a JWT ?

**Comment by stianst** [comment]
We have feature flags at the server level already, but not all that many features. Going forwards new features will mostly be wrapped as a feature.

We are also considering supporting setting features at the realm level. Not all features will support this, but some will. This will also brings some consistency as currently we have some options throughout the realm that does this kinda thing, but not in a consistent way. For example the new organizations can be enabled/disable on the server level through a feature flag, but at the realm level there is a separate organization option.

Not sure about client and user level, that does make sense, but only for some specific features. I think we'll focus on realm level first, and see where to go next.

With regard to unleash I don't think we should integrate directly with that, but we could do an SPI that allows creating custom logic to decide if a feature should be enabled or not.

**Reply by fosple** [reply]
I guess the feature request is for having Keycloak manage feature flags for other applications (similar to https://workos.com/docs/feature-flags ). Not for enabling features inside Keycloak. So one can setup feature flags, assign users to it and then implement the flag behaviour in their own applications.

**Reply by arsangamal** [reply]
> I guess the feature request is for having Keycloak manage feature flags for other applications (similar to https://workos.com/docs/feature-flags ). Not for enabling features inside Keycloak. So one can setup feature flags, assign users to it and then implement the flag behaviour in their own applications. 

Yes exactly, this is what I mean. So as you manage keycloak, you can manage the feature flags inside to control your app

**Reply by stianst** [reply]
I'd say providing an explicit feature flag feature like that would be out of scope for Keycloak; however, there are already easy ways you could achieve that with roles or user attributes.
